framework/contrib/hit/lex.h

```diff
@@ -0,0 +1,151 @@
+  std::vector<Token> tokens();
+  /// run executes the lexer causing it to processes the input and generate a sequence of tokens
+  /// representing it.  start is the first, starting LexFunc.  run calls start, passing in this
+  /// lexer instance as an argument.  If start returns another LexFunc, it is called with this lexer
+  /// instance as an argument, and so forth until the current LexFunc returns nullptr - at which
+  /// point lexing terminates.
+  std::vector<Token> run(LexFunc start);
+
+  /// emit finalizes and creates+adds a token of the given type to the list of tokens representing
+  /// the lexer's input text (i.e. the tokens returned by the tokens function).  The
+  /// created/emitted token is automatically populated with its value set to the string delineated by
+  /// the lexer's start and pos offsets into the input text.  emit then advanes the lexer start
+  /// offset to the current pos.
+  void emit(TokType type);
+
+  /// error emits an error token with the given messsage.  For convenience returns an nullptr
+  /// LexFunc that can be directly returns by any LexFunc that calls error.
+  LexFunc error(std::string msg);
+
+  /// next consumes and returns the next byte of input, advancing the lexer's position
+  /// offset by one.
+  char next();
+
+  /// accept conditionally accepts the next byte of input if it is one of the characters in the
+  /// valid set.  It returns true if a character was consumed and false otherwise.
+  bool accept(std::string valid);
+
+  /// acceptRun accepts a run of zero or more consecutive characters (greedily) that must be from
+  /// the given valid set.  It returns the number of characters consumed.
+  int acceptRun(std::string valid);
+
+  /// peek returns the next character of input without advancing the lexer's position offset (i.e.
+  /// without consuming the character).
+  char peek();
+
+  /// ignore advances the lexer's start offset to the current position offset without emitting any
+  /// tokens - effectively skipping a portion of the input text.
+  void ignore();
+  /// backup unconsumes the most recently consumed byte of input, reducing the position offset by
+  /// one.  This should only be called at most once after each call to next.
+  void backup();
+
+  /// input returns the full input text the lexer is operating on i.e. the entire input string it
+  /// was initialized/constructed with.
+  std::string input();
+  /// start returns the current start byte offset into the input identifying the start of the next
+  /// token that will be emitted (or next section of input that will be ignored).
+  size_t start();
+  /// pos returns the byte offset of the lexer's current position in the input text - i.e. the
+  /// location up to which input has been consumed/seen.
+  size_t pos();
+
+private:
+  std::string _name;
+  std::string _input;
+  size_t _start = 0;
+  size_t _pos = 0;
+  int _width = 0;
+  std::vector<Token> _tokens;
+};
+
+#define EOF (-1)
```

Comment:
I think the actual value is _implementation dependent_. You might want to save that into a different pp symbol, or just rename your enum entry...

Reply:
I know - I couldn't decide on a satisfactory solution, so the code just stayed like this :-/